### PR TITLE
[4.0] Permissions now custom element

### PR DESCRIPTION
**Pull request by anuragteapot**
Pull Request for Issue #19427 

### Summary of Changes

1. Move the code to the custom element
2. Fix the icon position

### Testing Instructions
1. Go to Admin Global Configuration and test the permissions
2. Change some of the permissions

### Expected result
Working as before

### Actual result
Working as before

### Documentation Changes Required
No

@dgt41 @C-Lodder  PR closed in <code>joomla-es6</code> repo. reopen in main repo.

### CODE REVIEW

**Comment by anuragteapot** [comment]
@astridx can you test this?
Thanks

**Comment by anuragteapot** [comment]
@franz-wohlkoenig can you test this ?

**Comment by astridx** [comment]
I have tested this item :white_check_mark: successfully on 88b63194680ad62f26adb648c242ee93c6ff7f98<br /><br />I applied this path and saw, that the permissions are now in the webcomponent <joomla-field-permissions ..>

I changed some permissions in global configuration. The changed permissions are saved and the inheritance to components works fine. For example, in the global configuration, I set the create permission to denied for the author. In the components com_content I see now that the create = allow is no longer inherited. And I was not able to apply the permisson.

I have created a new user group under registered. The permission have been set correctly (like registered) in the webcomponent <joomla-field-permissions ..>.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/19889">issues.joomla.org/tracker/joomla-cms/19889</a>.</sub>

**Comment by ghost** [comment]
@Anu1601CS i didn't test anymore as my lack of Knowledge generates the Risk of false positive Results.

**Comment by anuragteapot** [comment]
@franz-wohlkoenig Its ok

**Comment by astridx** [comment]
@franz-wohlkoenig I do not understand this. I think it's good to have a lot of tests and each test brings something possitive.
It is important that we write down what we have tested, so those who are merge can judge this.

**Comment by C-Lodder** [comment]
@franz-wohlkoenig 

1. Go to permissions view.
2. Select an option from the dropdown
3. A loading/spinner icon should appear whilst the request is being executed
4. The label/badge should change to red/green depending on the value selected
5. Refresh the page and ensure the value is the same just to make sure it was properly saved (sanity check)

**Comment by anuragteapot** [comment]
Dropped some extra files. Please test again.
Thanks

**Comment by anuragteapot** [comment]
@astridx @franz-wohlkoenig Can you please test again?
Thanks

**Comment by laoneo** [comment]
@dgt41 can you give here the final go to get it merged.

**Comment by dgrammatiko** [comment]
@laoneo go for it

File: media/system/webcomponents/js/joomla-field-permissions.js [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+				return pair[1];
+			}	
+		}
+		return false;
+	}		
+}
+
+customElements.define('joomla-field-permissions', JoomlaFieldPermissions);
+
+})(customElements, Joomla);
```
Comment (laoneo) [inline media/system/webcomponents/js/joomla-field-permissions.js:152]: New line is missing

File: media/system/webcomponents/js/joomla-field-permissions-es5.js [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+				}
+				return false;
+			}
+		}]);
+
+		return JoomlaFieldPermissions;
+	}(HTMLElement);
+
+	customElements.define('joomla-field-permissions', JoomlaFieldPermissions);
+})(customElements, Joomla);
```
Comment (laoneo) [inline media/system/webcomponents/js/joomla-field-permissions-es5.js:174]: New line is missing

File: media/system/js/fields/sendtestmail.min.js [hunk 3]
```diff
@@ -1 +0,0 @@
...ySelector('[name="jform[smtpauth]"]').value,smtpuser:document.querySelector('[name="jform[smtpuser]"]').value,smtppass:document.querySelector('[name="jform[smtppass]"]').value,smtphost:document.querySelector('[name="jform[smtphost]"]').value,smtpsecure:document.querySelector('[name="jform[smtpsecure]"]').value,smtpport:document.querySelector('[name="jform[smtpport]"]').value,mailfrom:document.querySelector('[name="jform[mailfrom]"]').value,fromname:document.querySelector('[name="jform[fromname]"]').value,mailer:document.querySelector('[name="jform[mailer]"]').value,mailonline:document.querySelector('[name="jform[mailonline]"]').value};Joomla.removeMessages(),Joomla.request({url:document.getElementById("sendtestmail").getAttribute("data-ajaxuri"),method:"POST",data:JSON.stringify(e),perform:!0,headers:{"Content-Type":"application/x-www-form-urlencoded"},onSuccess:function(e,t){"object"==typeof(e=JSON.parse(e)).messages&&null!==e.messages&&Joomla.renderMessages(e.messages)},onError:function(e){Joomla.renderMessages(Joomla.ajaxErrorsMessages(e))}})};document.addEventListener("DOMContentLoaded",function(){document.getElementById("sendtestmail").addEventListener("click",sendTestMail)});
```
Comment (laoneo) [inline media/system/js/fields/sendtestmail.min.js]: Why did you remove the content here?

Comment (anuragteapot) [reply media/system/js/fields/sendtestmail.min.js]: send test mail is also converted to our custom element see #19426.
But forgot to drop in that PR so I dropped this in my PR

File: media/system/webcomponents/js/joomla-field-permissions.js (hunk 1 above)
Comment (anuragteapot) [reply media/system/webcomponents/js/joomla-field-permissions.js:152]: Ok

File: media/system/webcomponents/js/joomla-field-permissions-es5.js (hunk 2 above)
Comment (dgrammatiko) [reply media/system/webcomponents/js/joomla-field-permissions-es5.js:174]: @laoneo no this file is autogenerated, the only js source here is in /biuld/webcomponents/....

File: media/system/webcomponents/js/joomla-field-permissions.js (hunk 1 above)
Comment (dgrammatiko) [reply media/system/webcomponents/js/joomla-field-permissions.js:152]: same comment as before, this is not the source file so it's ok as is

File: media/system/js/fields/sendtestmail.min.js (hunk 3 above)
Comment (laoneo) [reply media/system/js/fields/sendtestmail.min.js]: Please create for it a new pr and revert this one. Just to have things clearly separated. Thanks for understanding.

File: media/system/webcomponents/js/joomla-field-permissions-es5.js (hunk 2 above)
Comment (laoneo) [reply media/system/webcomponents/js/joomla-field-permissions-es5.js:174]: Would it be easy to fix the transpiler then?

File: media/system/js/fields/sendtestmail.min.js (hunk 3 above)
Comment (anuragteapot) [reply media/system/js/fields/sendtestmail.min.js]: Ok

File: media/system/webcomponents/js/joomla-field-permissions-es5.js (hunk 2 above)
Comment (dgrammatiko) [reply media/system/webcomponents/js/joomla-field-permissions-es5.js:174]: why, this is transpiled code we shouldn't care about the cs there

File: media/system/js/fields/sendtestmail.min.js (hunk 3 above)
Comment (anuragteapot) [reply media/system/js/fields/sendtestmail.min.js]: @laoneo should I mention this part in the description? Then it can be merged?
No, need to create extra PR for this.

Comment (laoneo) [reply media/system/js/fields/sendtestmail.min.js]: PR's should focus on one task. If we later have to revert it for some reasons, the invalid file is again there. So please can you revert that change.

Comment (anuragteapot) [reply media/system/js/fields/sendtestmail.min.js]: Ok

**Comment by anuragteapot** [comment]
@laoneo Done!

**Comment by laoneo** [comment]
Thanks. Good work.

**Comment by anuragteapot** [comment]
@laoneo Your welcome.
